endtoend/evaluators/metrics.go

```diff
@@ -0,0 +1,208 @@
+		}
+
+		genesis, err := eth.NewNodeClient(conns[i]).GetGenesis(context.Background(), &ptypes.Empty{})
+		if err != nil {
+			return err
+		}
+		forkDigest, err := p2putils.CreateForkDigest(time.Unix(genesis.GenesisTime.Seconds, 0), genesis.GenesisValidatorsRoot)
+		if err != nil {
+			return err
+		}
+
+		chainHead, err := eth.NewBeaconChainClient(conns[i]).GetChainHead(context.Background(), &ptypes.Empty{})
+		if err != nil {
+			return err
+		}
+		timeSlot, err := getValueOfTopic(pageContent, "beacon_clock_time_slot")
+		if err != nil {
+			return err
+		}
+		if chainHead.HeadSlot != uint64(timeSlot) {
+			return fmt.Errorf("expected metrics slot to equal chain head slot, expected %d, received %d", chainHead.HeadSlot, timeSlot)
+		}
+
+		for _, test := range metricLessThanTests {
+			topic := test.topic
+			if strings.Contains(topic, "%x") {
+				topic = fmt.Sprintf(topic, forkDigest)
+			}
+			if err := metricCheckLessThan(pageContent, topic, test.value); err != nil {
+				return errors.Wrapf(err, "failed %s check", test.name)
+			}
+		}
+		for _, test := range metricComparisonTests {
+			topic1 := test.topic1
+			if strings.Contains(topic1, "%x") {
+				topic1 = fmt.Sprintf(topic1, forkDigest)
+			}
+			topic2 := test.topic2
+			if strings.Contains(topic2, "%x") {
+				topic2 = fmt.Sprintf(topic2, forkDigest)
+			}
+			if err := metricCheckComparison(pageContent, topic1, topic2, test.expectedComparison); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func metricCheckLessThan(pageContent string, topic string, value int) error {
+	topicValue, err := getValueOfTopic(pageContent, topic)
+	if err != nil {
+		return err
+	}
+	if topicValue >= value {
+		return fmt.Errorf(
+			"unexpected result for metric %s, expected less than %d, received %d",
+			topic,
+			value,
+			topicValue,
+		)
+	}
+	return nil
+}
+
+func metricCheckComparison(pageContent string, topic1 string, topic2 string, comparison float64) error {
+	topic2Value, err := getValueOfTopic(pageContent, topic2)
+	if err != nil {
+		return err
+	}
+	topic1Value, err := getValueOfTopic(pageContent, topic1)
+	// If we can't find the first topic (error metrics), then assume the test passes.
+	if topic1Value == -1 && topic2Value != -1 {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	topicComparison := float64(topic1Value) / float64(topic2Value)
+	if topicComparison >= comparison {
+		return fmt.Errorf(
+			"unexpected result for comparison between metric %s and metric %s, expected comparison to be %.2f, received %.2f",
+			topic1,
+			topic2,
+			comparison,
+			topicComparison,
+		)
+	}
+	return nil
+}
+
+func getValueOfTopic(pageContent string, topic string) (int, error) {
+	// Adding a space to search exactly.
+	startIdx := strings.LastIndex(pageContent, topic+" ")
+	if startIdx == -1 {
+		return -1, fmt.Errorf("did not find requested text %s in %s", topic, pageContent)
+	}
+	endOfTopic := startIdx + len(topic)
+	fmt.Println(pageContent[startIdx:endOfTopic])
```

Comment:
is `fmt.Println` here intended or just for debug?

Reply:
My bad, thanks.